File: test_undo.py
import json

tpvState = {
    "lineas": [],
    "historyStack": [],
    "historyIndex": 0
}

def saveHistoryState(isInitial=False):
    # Sanitize
    stateParaGuardar = []
    for l in tpvState["lineas"]:
        resto = l.copy()
        if "id" in resto:
            del resto["id"]
        stateParaGuardar.append(resto)
        
    newStateStr = json.dumps(stateParaGuardar, separators=(',', ':'))
    print("Saving:", newStateStr)
    
    if isInitial:
        tpvState["historyStack"] = [newStateStr]
        tpvState["historyIndex"] = 0
    else:
        if len(tpvState["historyStack"]) > 0 and tpvState["historyStack"][tpvState["historyIndex"]] == newStateStr:
            print("DUPLICATE, skipping.")
            return

        if tpvState["historyIndex"] < len(tpvState["historyStack"]) - 1:
            tpvState["historyStack"] = tpvState["historyStack"][:tpvState["historyIndex"] + 1]

        tpvState["historyStack"].append(newStateStr)
        tpvState["historyIndex"] += 1

def print_state():
    print(f"Index: {tpvState['historyIndex']}, Stack lengths: {[len(x) for x in tpvState['historyStack']]}")

print("INITIAL")
saveHistoryState(True)
print_state()

print("\nCLICK 1")
tpvState["lineas"].append({"_uid": "l_1", "cantidad": 1})
saveHistoryState()
print_state()

print("\nCLICK 2")
tpvState["lineas"][0]["cantidad"] = 2
saveHistoryState()
print_state()

print("\nCLICK 3")
tpvState["lineas"][0]["cantidad"] = 3
saveHistoryState()
print_state()

print("\nCLICK 4")
tpvState["lineas"][0]["cantidad"] = 4
saveHistoryState()
print_state()

print("\nCLICK 5")
tpvState["lineas"][0]["cantidad"] = 5
saveHistoryState()
print_state()
print("Visible lineas:", tpvState["lineas"])

print("\nCLICK UNDO 1")
tpvState["historyIndex"] -= 1
tpvState["lineas"] = json.loads(tpvState["historyStack"][tpvState["historyIndex"]])
print_state()
print("Visible lineas:", tpvState["lineas"])

print("\nCLICK UNDO 2")
tpvState["historyIndex"] -= 1
tpvState["lineas"] = json.loads(tpvState["historyStack"][tpvState["historyIndex"]])
print_state()
print("Visible lineas:", tpvState["lineas"])

print("\nCLICK REDO 1")
tpvState["historyIndex"] += 1
tpvState["lineas"] = json.loads(tpvState["historyStack"][tpvState["historyIndex"]])
print_state()
print("Visible lineas:", tpvState["lineas"])

File: test_undo.js
const tpvState = {
    lineas: [],
    historyStack: [],
    historyIndex: 0
};

function saveHistoryState(isInitial = false) {
    const stateParaGuardar = tpvState.lineas.map(l => {
        const { id, ...resto } = l;
        return resto;
    });

    const newStateStr = JSON.stringify(stateParaGuardar);
    console.log("Saving:", newStateStr);

    if (isInitial) {
        tpvState.historyStack = [newStateStr];
        tpvState.historyIndex = 0;
    } else {
        if (
            tpvState.historyStack.length > 0 &&
            tpvState.historyStack[tpvState.historyIndex] === newStateStr
        ) {
            console.log("DUPLICATE, skipping.");
            return;
        }

        if (tpvState.historyIndex < tpvState.historyStack.length - 1) {
            tpvState.historyStack = tpvState.historyStack.slice(0, tpvState.historyIndex + 1);
        }

        tpvState.historyStack.push(newStateStr);
        tpvState.historyIndex++;
    }
}

function printState() {
    console.log(`Index: ${tpvState.historyIndex}, Stack:`, tpvState.historyStack);
}

console.log("INITIAL");
saveHistoryState(true);
printState();

console.log("\nCLICK 1");
tpvState.lineas.push({ _uid: 'l_1', cantidad: 1 });
saveHistoryState();
printState();

console.log("\nCLICK 2");
tpvState.lineas[0].cantidad = 2;
saveHistoryState();
printState();

console.log("\nCLICK 3");
tpvState.lineas[0].cantidad = 3;
saveHistoryState();
printState();

console.log("\nCLICK UNDO 1");
tpvState.historyIndex--;
tpvState.lineas = JSON.parse(tpvState.historyStack[tpvState.historyIndex]);
printState();
console.log("Current lineas view:", tpvState.lineas);

console.log("\nCLICK UNDO 2");
tpvState.historyIndex--;
tpvState.lineas = JSON.parse(tpvState.historyStack[tpvState.historyIndex]);
printState();
console.log("Current lineas view:", tpvState.lineas);

console.log("\nCLICK REDO 1");
tpvState.historyIndex++;
tpvState.lineas = JSON.parse(tpvState.historyStack[tpvState.historyIndex]);
printState();
console.log("Current lineas view:", tpvState.lineas);
